parsers/json/reader.cc

```diff
@@ -0,0 +1,149 @@
+#include "json.h"
+#include "trieste/source.h"
+
+namespace trieste::json
+{
+  JSONReader::JSONReader(const std::filesystem::path& path)
+  : JSONReader(SourceDef::load(path))
+  {}
+
+  JSONReader::JSONReader(const std::string& yaml)
+  : JSONReader(SourceDef::synthetic(yaml))
+  {}
+
+  JSONReader::JSONReader(const Source& source)
+  : m_source(source),
+    m_debug_enabled(false),
+    m_debug_path("."),
+    m_well_formed_checks_enabled(false)
+  {}
+
+  JSONReader& JSONReader::debug_enabled(bool value)
+  {
+    m_debug_enabled = value;
+    return *this;
+  }
+
+  bool JSONReader::debug_enabled() const
+  {
+    return m_debug_enabled;
+  }
+
+  JSONReader& JSONReader::well_formed_checks_enabled(bool value)
+  {
+    m_well_formed_checks_enabled = value;
+    return *this;
+  }
+
+  bool JSONReader::well_formed_checks_enabled() const
+  {
+    return m_well_formed_checks_enabled;
+  }
+
+  JSONReader& JSONReader::debug_path(const std::filesystem::path& path)
+  {
+    m_debug_path = path;
+    return *this;
+  }
+
+  const std::filesystem::path& JSONReader::debug_path() const
+  {
+    return m_debug_path;
+  }
+
+  Node has_parse_error(Node n)
+  {
+    std::vector<Node> stack;
+    stack.push_back(n);
+    while (!stack.empty())
+    {
+      auto current = stack.back();
+      stack.pop_back();
+      if (current->in({Error, ErrorSeq}))
+      {
+        return current;
+      }
+      for (auto& child : *current)
+      {
+        stack.push_back(child);
+      }
+    }
+
+    return nullptr;
+  }
+
+  void JSONReader::read()
+  {
+    auto ast = NodeDef::create(Top);
+    Parse parse = json::parser();
+    ast << parse.sub_parse("json", File, m_source);
+    Node maybe_error = has_parse_error(ast);
+    if (has_parse_error(ast))
```

Comment:
You're calling `has_parse_error` twice.  Is this correct?

Reply:
Nope that's a bug.